### PR TITLE
Add validation for declared volume claim templates

### PR DESCRIPTION
**Pull request by pebrc**
Partially addresses https://github.com/elastic/cloud-on-k8s/issues/4525

Adds a validation to check that declared volume claim templates are actually mounted in any of the containers. This validation does not apply if the user configures the documented default data volume `elasticsearch-data` for which ECK manages the volume mount. 

There is still a possibility for misconfiguration. As we are currently supporting custom volume mounts and custom `data.path` settings in Elasticsearch it is quite hard to fully validate the configuration. With the proposed validation it would still be possible for a user to configure a custom volume claim template, mount it in a side car and have no persistent Elasticsearch data volume. This could be addressed in two ways:
* also parse the Elasticsearch configuration during validation and make sure each `data.path` has a volume mount
* always configure a default data volume as we had it pre-1.4.0

### CODE REVIEW

File: pkg/controller/elasticsearch/validation/validations.go [hunk 1]
```diff
@@ -34,6 +34,7 @@ const (
 	parseStoredVersionErrMsg = "Cannot parse current Elasticsearch version. String format must be {major}.{minor}.{patch}[-{label}]"
 	parseVersionErrMsg       = "Cannot parse Elasticsearch version. String format must be {major}.{minor}.{patch}[-{label}]"
 	pvcImmutableErrMsg       = "volume claim templates can only have their storage requests increased, if the storage class allows volume expansion. Any other change is forbidden"
+	pvcNotMounted            = "volume claim declared but volume not mounted in any container"
```
Comment (sebgl) [inline pkg/controller/elasticsearch/validation/validations.go:37]: Should we make it more explicit to cover the usual user mistake?
`volume claim declared but volume not mounted in any container. Please note the Elasticsearch data volume should be named "elasticsearch-data"`

Comment (sebgl) [inline pkg/controller/elasticsearch/validation/validations.go:37]: `pvcNotMountedErrMsg` for consistency?

**Review by sebgl — APPROVED** [review]
LGTM. Left a few comments for minor improvements.

One concern is that existing users with misconfigured volumes would suddenly have their cluster reconciliation error-out when they upgrade the operator to a version with that validation. I think that's OK (better point out the problem and stop reconciling rather than silently ignoring it?).

> With the proposed validation it would still be possible for a user to configure a custom volume claim template, mount it in a side car and have no persistent Elasticsearch data volume. This could be addressed in two ways

I think I'm fine with not addressing this edge case.

**Comment by pebrc** [comment]
Jenkins test this please

**Comment by aliciascott** [comment]
Happened again here https://github.com/elastic/sdh-elasticsearch/issues/4741
